### PR TITLE
Add integration with Hugging Face

### PR DESCRIPTION
**Pull request by NielsRogge**
Hi,

Thanks for this nice work. I wrote a quick PoC to showcase that you can easily have integration so that you can automatically load the various AIM models using `from_pretrained` (and push them using `push_to_hub`), track download numbers for your models (similar to models in the Transformers library), and have nice model cards on a per-model basis. It leverages the [PyTorchModelHubMixin](https://huggingface.co/docs/huggingface_hub/package_reference/mixins#huggingface_hub.PyTorchModelHubMixin) class which allows to inherits these methods.

Usage is as follows:

```
from aim.models import AIMForImageClassification
from aim.torch.data import val_transforms
from PIL import Image

img = Image.open(...)
model = AIMForImageClassification.from_pretrained("apple/aim-600m")
transform = val_transforms()

inp = transform(img).unsqueeze(0)
logits, _ = model(inp)
```
The corresponding model is here for now: https://huggingface.co/nielsr/aim-600m. We could move all checkpoints to separate repos on the [Apple organization](https://huggingface.co/apple) if you're interested.

I saw that currently download numbers won't be tracked since all checkpoints are just part of a single repository.

Would you be interested in this integration?

Kind regards,

Niels

### CODE REVIEW

**Comment by aelnouby** [comment]
Thanks @NielsRogge, the plan sounds good. @michalk8 and I will provide a review for this PR shortly.

File: src/aim/pytorch/hub_mixin.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+"""Illustrates the mixin for one class."""
```
Comment (michalk8) [inline src/aim/pytorch/hub_mixin.py:1]: In general, all files should start with a license, please add.

File: src/aim/pytorch/__init__.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
 # For licensing see accompanying LICENSE file.
 # Copyright (C) 2024 Apple Inc. All Rights Reserved.
+from aim import mixins
```
Comment (michalk8) [inline src/aim/pytorch/__init__.py:3]: There should be no need for this, would remove.

File: src/aim/pytorch/hub_mixin.py (hunk 1 above)
Comment (michalk8) [inline src/aim/pytorch/hub_mixin.py:1]: I think this example can be either added to `README.md` or let's create `examples` folder. In any case, I don't think it should be placed inside the package itself.

File: src/aim/pytorch/models.py [hunk 3]
```diff
@@ -82,7 +84,91 @@ def forward(
         return tokens, features
 
 
-class AIM(mixins.AIMMixin, nn.Module):
+class AttrDict(dict):
```
Comment (michalk8) [inline src/aim/pytorch/models.py:87]: I'd add this to `aim/utils.py`

File: src/aim/pytorch/models.py [hunk 4]
```diff
@@ -82,7 +84,91 @@ def forward(
...
+        max_block_id: Optional[int] = -1,
+    ) -> List[ArrayLike]:
+        x = self.preprocessor(x, mask=mask)
+        feats = self.trunk(
+            x, mask=mask, max_block_id=max_block_id, return_features=True
+        )
+        return feats
+
+
+class AIM(nn.Module):
```
Comment (michalk8) [inline src/aim/pytorch/models.py:171]: This should still inherit the mixin, otherwise it breaks the API.

File: src/aim/pytorch/models.py [hunk 5]
```diff
@@ -82,7 +84,91 @@ def forward(
...
 
 
-class AIM(mixins.AIMMixin, nn.Module):
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self    
+
+
+class AIMForImageClassification(nn.Module, PyTorchModelHubMixin):
```
Comment (michalk8) [inline src/aim/pytorch/models.py:93]: Unless I'm missing something, this should work without re-defining the `forward`, etc.:
```python
class AIMForImageClassification(mixins.AIMMixin, PyTorchModelHubMixin, nn.Module):
    ...
```

File: src/aim/pytorch/models.py [hunk 6]
```diff
@@ -3,14 +3,16 @@
 # --------------------------------------------------
 # References:
 # https://github.com/facebookresearch/ImageBind
-from typing import Any, Callable, List, Optional, Tuple, Union
+from typing import Any, Callable, List, Optional, Tuple, Union, Dict
 
 import torch
 from torch import nn
 
-from .. import mixins
```
Comment (michalk8) [inline src/aim/pytorch/models.py:11]: Is there a reason for changing `from . import layers` to `import layers`? If not, prefer the latter.

File: src/aim/pytorch/models.py [hunk 7]
```diff
@@ -3,14 +3,16 @@
...
-from . import layers
+import layers
 
+from huggingface_hub import PyTorchModelHubMixin
+
+ArrayLike = Any
```
Comment (michalk8) [inline src/aim/pytorch/models.py:15]: After using the mixin class, this should not be necessary.

File: src/aim/pytorch/models.py (hunk 5 above)
Comment (michalk8) [reply src/aim/pytorch/models.py:93]: Also please don't forget to add the `AIMForImageClassification` to `__all__`

File: src/aim/pytorch/models.py [hunk 8]
```diff
@@ -82,7 +84,91 @@ def forward(
...
 
-class AIM(mixins.AIMMixin, nn.Module):
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self    
+
+
+class AIMForImageClassification(nn.Module, PyTorchModelHubMixin):
+    def __init__(self, config: dict):
```
Comment (michalk8) [inline src/aim/pytorch/models.py:94]: config: Dict[str, Any]

File: src/aim/pytorch/hub_mixin.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+import torch
```
Comment (michalk8) [inline src/aim/pytorch/hub_mixin.py:3]: Please install and run `pre-commit` for formatting.

**Review by michalk8 — CHANGES_REQUESTED** [review]
Thanks a lot @NielsRogge , we'd be happy for your contribution!
Note that we switched from `src/aim` layout to just flat layout (`aim/`).

**Comment by kohlamor** [comment]
Check my repository [Apple Books]

**Pull request closed without merging**